src/core/composer/qgscomposerpointsbasedshape.cpp

```diff
@@ -0,0 +1,404 @@
+  double max_distance = ( checkArea ) ? radius : min_distance;
+  bool rc = false;
+  int idx = -1;
+
+  for ( int i = 0; i != mPolygon.size(); i++ )
+  {
+    // get points of polyline
+    const QPointF pt1 = mPolygon[i];
+    QPointF pt2 = mPolygon[0];
+    if (( i + 1 ) != mPolygon.size() )
+      pt2 = mPolygon[i+1];
+
+    // compute line eq
+    const double coef = ( pt2.y() - pt1.y() ) / ( pt2.x() - pt1.x() );
+    const double b = pt1.y() - coef * pt1.x();
+
+    double distance = std::numeric_limits<double>::max();
+    if ( isinf( coef ) )
+      distance = qAbs( pt1.x() - start.x() );
+    else
+    {
+      const double coef2 = ( -1 / coef );
+      const double b2 = start.y() - coef2 * start.x();
+
+      QPointF inter;
+      if ( isinf( coef2 ) )
+      {
+        distance = qAbs( pt1.y() - start.y() );
+        inter.setX( start.x() );
+        inter.setY( pt1.y() );
+      }
+      else
+      {
+        const double interx = ( b - b2 ) / ( coef2 - coef );
+        const double intery = interx * coef2 + b2;
+        inter.setX( interx );
+        inter.setY( intery );
+      }
+
+      // check if intersection is within the line
+      const double length1 = computeDistance( inter, pt1 );
+      const double length2 = computeDistance( inter, pt2 );
+      const double length3 = computeDistance( pt1, pt2 );
+      const double length4 = length1 + length2;
+
+      if ( qAbs( length3 - length4 ) < std::numeric_limits<float>::epsilon() )
+        distance = computeDistance( inter, start );
+    }
+
+    if ( distance < min_distance && distance < max_distance )
+    {
+      min_distance = distance;
+      idx = i;
+    }
+  }
+
+  if ( idx >= 0 )
+  {
+    rc = _addPoint( idx, start, max_distance );
+    updateSceneRect();
+  }
+
+  return rc;
+}
+
+QPointF QgsComposerPointsBasedShape::convertToItemCoordinate( const QPointF &point )
+{
+  QPointF pt;
+  pt.setX( point.x() - scenePos().x() );
+  pt.setY( point.y() - scenePos().y() );
+
+  return pt;
+}
+
+void QgsComposerPointsBasedShape::drawPoints( QPainter *painter ) const
+{
+  double rectSize = 3.0 / horizontalViewScaleFactor();
+
+  QgsStringMap properties;
+  properties.insert( "name", "cross" );
+  properties.insert( "color_border", "red" );
+
+  QScopedPointer<QgsMarkerSymbolV2> symbol;
+  symbol.reset( QgsMarkerSymbolV2::createSimple( properties ) );
+  symbol.data()->setSize( rectSize );
+  symbol.data()->setAngle( 45 );
+
+  QgsMapSettings ms = mComposition->mapSettings();
+  ms.setOutputDpi( painter->device()->logicalDpiX() );
+
+  QgsRenderContext context = QgsRenderContext::fromMapSettings( ms );
+  context.setPainter( painter );
+  context.setForceVectorOutput( true );
+
+  QScopedPointer<QgsExpressionContext> expressionContext;
+  expressionContext.reset( createExpressionContext() );
+  context.setExpressionContext( *expressionContext.data() );
+
+  symbol.data()->startRender( context );
+  QVector<QPointF>::const_iterator itnode = mPolygon.begin();
```

Comment:
Better to go:

```
QVector<QPointF>::const_iterator itnode = mPolygon.constBegin();
for ( ; itnode != mPolygon.constEnd(); ++itnode )

```

Using constBegin/constEnd for const iterators can avoid a potential detachment.